src/mlio/image_reader.cxx

```diff
@@ -0,0 +1,392 @@
+        throw std::invalid_argument{fmt::format(
+            "Invalid value for the resize operation: {0:n}. Value should be "
+            "greater than 0",
+            params_.resize)};
+    }
+
+    int new_height, new_width;
+    if (src.rows > src.cols) {
+        new_height = params_.resize * src.rows / src.cols;
+        new_width = params_.resize;
+    }
+    else {
+        new_height = params_.resize;
+        new_width = params_.resize * src.cols / src.rows;
+    }
+    try {
+        cv::resize(src, dst, cv::Size(new_width, new_height), 0, 0);
+    }
+    catch (cv::Exception &e) {
+        if (bbh_ != bad_batch_handling::skip) {
+            auto msg =
+                fmt::format("Image resize operation failed: {0:s}", e.what());
+            if (bbh_ == bad_batch_handling::error) {
+                throw std::runtime_error{msg};
+            }
+            logger::warn(msg);
+        }
+        skip_batch = true;
+    }
+}
+
+void
+image_reader::crop(cv::Mat &src,
+                   cv::Mat &dst,
+                   std::atomic_bool &skip_batch) const
+{
+    if (src.rows < params_.image_dimensions[1] ||
+        src.cols < params_.image_dimensions[2]) {
+        if (bbh_ != bad_batch_handling::skip) {
+            auto msg = fmt::format(
+                "Input image [rows: {0:n}, cols: {1:n}] is smaller than "
+                "the output image image_dimensions [rows: {2:n}, cols: {3:n}]",
+                src.rows,
+                src.cols,
+                params_.image_dimensions[1],
+                params_.image_dimensions[2]);
+            if (bbh_ == bad_batch_handling::error) {
+                throw std::invalid_argument{msg};
+            }
+            logger::warn(msg);
+        }
+        skip_batch = true;
+        return;
+    }
+
+    // center crop
+    int y = (src.rows - params_.image_dimensions[1]) / 2;
+    int x = (src.cols - params_.image_dimensions[2]) / 2;
+
+    try {
+        cv::Rect roi(
+            x, y, params_.image_dimensions[2], params_.image_dimensions[1]);
+        src(roi).copyTo(dst);
+    }
+    catch (cv::Exception &e) {
+        if (bbh_ != bad_batch_handling::skip) {
+            auto msg =
+                fmt::format("Image crop operation failed: {0:s}", e.what());
+            if (bbh_ == bad_batch_handling::error) {
+                throw std::runtime_error{msg};
+            }
+            logger::warn(msg);
+        }
+        skip_batch = true;
+    }
+}
+
+}  // namespace v1
+}  // namespace mlio
+
+#else
+
+// clang-format off
+
+#include "mlio/not_supported_error.h"
+
+namespace cv {
+class Mat {};
+}  // namespace cv
+
+namespace mlio {
+inline namespace v1 {
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wmissing-noreturn"
+
+image_reader::image_reader(data_reader_params rdr_prm,
+                           image_reader_params image_prm)
+    : parallel_data_reader{std::move(rdr_prm)}, params_{std::move(image_prm)}
+{
+    bbh_ = {};
```

Comment:
Move this to the constructor initializer list, next to `params_` in line 325.

Reply:
Moving it there gives an error saying: `warning: private field 'img_dims_' is not used [clang-diagnostic-unused-private-field]`